src/librustc/infer/error_reporting/anon_anon_conflict.rs

```diff
@@ -27,66 +27,97 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
     //    { x.push(y); }.
     // The example gives
     // fn foo(x: &mut Vec<&u8>, y: &u8) {
-    //                    ---      --- these references must have the same lifetime
+    //                    ---      --- these references are declared with different lifetimes...
     //            x.push(y);
-    //            ^ data from `y` flows into `x` here
-    // It will later be extended to trait objects and structs.
+    //            ^ ...but data from `y` flows into `x` here
+    // It has been extended for the case of structs too.
+    // Consider the example
+    // struct Ref<'a> { x: &'a u32 }
+    // fn foo(mut x: Vec<Ref>, y: Ref) {
+    //                   ---      --- these structs are declared with different lifetimes...
+    //               x.push(y);
+    //               ^ ...but data from `y` flows into `x` here
+    // }
+    // It will later be extended to trait objects.
     pub fn try_report_anon_anon_conflict(&self, error: &RegionResolutionError<'tcx>) -> bool {
-
         let (span, sub, sup) = match *error {
             ConcreteFailure(ref origin, sub, sup) => (origin.span(), sub, sup),
             _ => return false, // inapplicable
         };
 
         // Determine whether the sub and sup consist of both anonymous (elided) regions.
-        let (ty1, ty2) = if self.is_suitable_anonymous_region(sup).is_some() &&
-                            self.is_suitable_anonymous_region(sub).is_some() {
+        let (ty1, ty2, scope_def_id_1, scope_def_id_2, bregion1, bregion2) = if
+            self.is_suitable_anonymous_region(sup, true).is_some() &&
+            self.is_suitable_anonymous_region(sub, true).is_some() {
             if let (Some(anon_reg1), Some(anon_reg2)) =
-                (self.is_suitable_anonymous_region(sup), self.is_suitable_anonymous_region(sub)) {
-                let ((_, br1), (_, br2)) = (anon_reg1, anon_reg2);
-                if self.find_anon_type(sup, &br1).is_some() &&
-                   self.find_anon_type(sub, &br2).is_some() {
-                    (self.find_anon_type(sup, &br1).unwrap(),
-                     self.find_anon_type(sub, &br2).unwrap())
-                } else {
-                    return false;
+                (self.is_suitable_anonymous_region(sup, true),
+                 self.is_suitable_anonymous_region(sub, true)) {
+                let ((def_id1, br1), (def_id2, br2)) = (anon_reg1, anon_reg2);
+                let found_arg1 = self.find_anon_type(sup, &br1);
+                let found_arg2 = self.find_anon_type(sub, &br2);
+                match (found_arg1, found_arg2) {
```

Comment:
It's more idiomatic to use `if let` instead of `match` here. Also, `found_arg_sup` >> `found_arg` for variable names, etc.